[Feature:API] Get list of student's gradeables

IDzyre opened the pull request:
### Why is this Change Important & Necessary?
The Submitty API is lacking in features that would allow for creating a VS code extension. Specifically, getting a list of gradeables that a user has access to for a given course, and a route to verify the user logged in.

Also, the current "get gradeable values" route has the full grade passed, but not the reason for each testcase passing/failing.

### What is the New Behavior?
A route has been added `base_url/api/<term>/<course>/gradeables` that returns a GradeableList converted to Json of the currently available gradeables to the given user.
An instructor can also use this, and specify a given user ID if they would like to get gradeables for a student.

A lot more information has been added to the `.../values` endpoint to give the extension more to work with.

### What steps should a reviewer take to reproduce or test the bug or new feature?
Review the code, or use an API testing tool (such as Postman) to test the API route successfully returns values for a given user.

### Other information
This won't impact current installations or require any sysadmin actions.

IDzyre committed:
Add gradeables list API route

IDzyre committed:
Change values added

Commit:
Merge main and fix merge conflicts

IDzyre committed:
Merge remote-tracking branch 'origin/main' into gradeables-list

IDzyre committed:
Merge remote-tracking branch 'origin/main' into gradeables-list

IDzyre committed:
Re-add self-hosted submissions

IDzyre committed:
Some linting

IDzyre committed:
Merge branch 'gradeables-list' of github.com:Submitty/Submitty into gradeables-list

IDzyre committed:
Add testcase checks

IDzyre committed:
Merge remote-tracking branch 'origin/testcase-checks-api' into gradeables-list

IDzyre committed:
Changes

IDzyre committed:
Merge branch 'gradeables-list' of github.com:Submitty/Submitty into gradeables-list

IDzyre committed:
Instructors can do anyone, don't need to specify user_id

IDzyre committed:
add gradaeble type, and VCS repository

IDzyre committed:
Add 'Me' route

IDzyre committed:
Rename function to not copy

IDzyre committed:
Put inside of array not just values

IDzyre committed:
Set to return Json response

IDzyre committed:
Merge branch 'main' into gradeables-list

IDzyre committed:
Merge branch 'main' into gradeables-list

IDzyre committed:
Use Base URL if partial path and base url are the same

IDzyre committed:
Merge branch 'main' into gradeables-list

IDzyre committed:
Move gradeables-list route, and fix lint

IDzyre committed:
Merge branch 'gradeables-list' of github.com:Submitty/Submitty into gradeables-list

IDzyre committed:
Fix php linting

IDzyre committed:
Linting, remove whitespace change, and remove redundant 'is_student'

IDzyre committed:
Use gradeable ID as key for array

IDzyre committed:
Merge branch 'main' into gradeables-list

IDzyre committed:
Linting, and change route to remote /list

IDzyre committed:
Don't require user_ids

IDzyre committed:
Try to fix duplicated URL

IDzyre committed:
Testing duplicated URL

IDzyre committed:
Removing testing, it needed rtrim

IDzyre committed:
Don't show vcs repository if not vcs, and add cypress api tests

IDzyre committed:
Merge branch 'main' into gradeables-list

IDzyre committed:
Fix linting

IDzyre committed:
Merge branch 'gradeables-list' of github.com:Submitty/Submitty into gradeables-list

IDzyre committed:
Remove port from api tests

IDzyre committed:
Merge branch 'main' into gradeables-list

IDzyre committed:
Merge branch 'main' into gradeables-list

IDzyre committed:
Merge branch 'main' into gradeables-list

IDzyre committed:
Update namespace reference for User group check

IDzyre committed:
Merge branch 'main' into gradeables-list

IDzyre committed:
Merge branch 'main' into gradeables-list

IDzyre committed:
Merge branch 'main' into gradeables-list

williamjallen commented on site/app/controllers/student/SubmissionController.php:
```diff
@@ -2590,4 +2600,25 @@ public function generateRepo(string $gradeable_id): RedirectResponse {
+            if ($user_id !== '') {
+                $user = $this->core->getQueries()->getUserById($user_id);
+            }
+        }
```
This feels risky.  Is this necessary?

IDzyre replied:
Not at this point, if it is to be added in the future, it should probably have stricter permissions checks.

williamjallen commented on site/app/controllers/UserProfileController.php:
```diff
@@ -45,6 +45,20 @@ public function showUserProfile(): MultiResponse {
+     *
+     */
+    #[Route("/api/me", methods: ["GET"])]
+    public function apiMeResponse(): JsonResponse {
+        $user = $this->core->getUser();
+        return JsonResponse::getSuccessResponse([
+            'user_id' => $user->getId(),
+            'user_given_name' => $user->getDisplayedGivenName(),
+            'user_family_name' => $user->getDisplayedFamilyName()
+        ]);
+    }
```
We can assume that the user exists here, right?

IDzyre replied:
I believe so, if I am not mistaken the router redirects routes to "unauthenticated access please login" page if the user isn't logged in.

williamjallen commented on site/app/models/gradeable/Gradeable.php:
```diff
@@ -2262,6 +2262,9 @@ public function getRepositoryPath(User $user, ?Team $team = null) {
         if (strpos($this->getVcsPartialPath(), '://') !== false || substr($this->getVcsPartialPath(), 0, 1) === '/') {
             $vcs_path = $this->getVcsPartialPath();
         }
+        elseif (rtrim($this->core->getConfig()->getVcsBaseUrl(), '/') === rtrim($this->getVcsPartialPath(), '/')) {
+                $vcs_path = rtrim($this->core->getConfig()->getVcsBaseUrl(), '/');
+        }
```
This should be reformatted.

IDzyre replied:
How do you suggest this should be formatted ?

williamjallen replied:
Line 2266 should be indented with 4 spaces.

IDzyre committed:
Remove instructor access for other users from gradeable list

This can be added back later if requested, but to get this merged it is being removed.

IDzyre committed:
Merge branch 'main' into gradeables-list

IDzyre committed:
Fix formatting and indenting

williamjallen reviewed (approved):
Looks good to me.  Let's wait to merge this until the CI is back to passing unless it's really blocking something else you're working on.

IDzyre commented:
It is what the VS Code plugin repo is based on, so the sooner the better. I will see if I can fix (at least a bulk) of CI tonight, unless someone else fixes it first.

bmcutler committed:
Merge branch 'main' into gradeables-list

IDzyre committed:
Merge branch 'main' into gradeables-list
